### PR TITLE
When emitting an arrow function, parenthesize the body if it could be interpreted as a block instead of an object literal.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/Microsoft/TypeScript/issues/2250

### CODE REVIEW

**Comment by JsonFreeman** [comment]
:+1:

File: src/compiler/emitter.ts [hunk 1]
```diff
@@ -4202,9 +4202,19 @@ module ts {
...
+                // For es6 and higher we can emit the expression as is.  However, in the case 
+                // where the expression might end up looking like a block when emitted, we'll
+                // also wrap it in parentheses first.  For example if you have: a => <foo>{}
+                // then we need to generate: a => ({})
                 write(" ");
-                emit(body);
+
+                // Unwrap all type assertions.
+                var current = body;
+                while (current.kind === SyntaxKind.TypeAssertionExpression) {
```
Comment (DanielRosenwasser) [inline src/compiler/emitter.ts]: What about when it's a type assertion in a parenthesized expression?

Comment (CyrusNajmabadi) [inline src/compiler/emitter.ts]: I'll add a test for that as well.

**Comment by DanielRosenwasser** [comment]
Add the following tests

```TypeScript
var a = () => <Error>{ name: "foo", message: "bar" };      
                                                           
var b = () => (<Error>{ name: "foo", message: "bar" });    
                                                           
var c = () => ({ name: "foo", message: "bar" });           
                                                           
var d = () => ((<Error>({ name: "foo", message: "bar" })));
```

**Pull request merged**